rmf_fleet_adapter/src/rmf_fleet_adapter/agv/FleetUpdateHandle.cpp

```diff
@@ -718,23 +635,50 @@ void FleetUpdateHandle::Implementation::dispatch_request_cb(
 auto FleetUpdateHandle::Implementation::is_valid_assignments(
   Assignments& assignments) const -> bool
 {
-  std::unordered_set<std::string> executed_tasks;
-  for (const auto& [context, mgr] : task_managers)
-  {
-    const auto& tasks = mgr->get_executed_tasks();
-    executed_tasks.insert(tasks.begin(), tasks.end());
-  }
-
+  std::size_t assignment_size = 0;
+  std::size_t queued_task_size = 0;
   for (const auto& agent : assignments)
   {
     for (const auto& a : agent)
     {
-      if (executed_tasks.find(a.request()->id()) != executed_tasks.end())
+      // If ID exists doesnt exist in queue_tasks
+      if (!is_queued(a.request()->id()))
         return false;
+      assignment_size++;
     }
   }
 
-  return true;
+  // Also check if size of total queued tasks is the same as assignments
+  for (const auto& mgr : task_managers)
+    queued_task_size += mgr->task_queue().size();
+
+  return (queued_task_size == assignment_size);
+}
+
+//==============================================================================
+bool FleetUpdateHandle::Implementation::is_queued(
+  const std::string& task_id ) const
+{
+  std::unordered_set<std::string> queued_tasks;
```

Comment:
Constructing a complete set of `queued_tasks` and then returning ```return (queued_tasks.find(task_id) != queued_tasks.end());``` is a bit inefficient. This way in the best case the function returns a lot earlier.

Instead you can have 

```
for (const auto& [context, mgr] : task_managers)
{
  for (const auto& request : mgr->requests())
  {
    if (request->id() == task_id)
      return True
  }
}
return false
```